### ISSUE TITLE
zig cc performance anomaly with relative ZIG_LIB_DIR or symlinks

### ISSUE THREAD

**Issue by motiejus** [ziglang/zig#13619]
This issue was rewritten after I narrowed down the problem with a small test case.

**Preparation**

**main.c**
```
int main() {
    return 0;
}
```

**repro.sh**

```bash
#!/bin/bash
set -euo pipefail

# INPUT: set this to the path to your zig directory. Should have $ORIG/lib.
readonly ORIG=/code/zig/stage3

readonly ZIGCC=(s/bin/zig cc -target x86_64-linux-gnu.2.28)

_time() {
    /usr/bin/time --format "  + %C\n  Wall time: %e" "$@"
}

echo -n "Cleaning up and preparing directories... "
rm -fr s ~/.cache/zig cache_zig_step*
cp -sr $ORIG s
echo done

echo -n "Warming up zig cache... "
echo | ${ZIGCC[@]} /dev/stdin >& /dev/null || :
echo done

echo "Plain build: expecting a quick build"
_time ${ZIGCC[@]} -DX=1 main.c -o /tmp/main1
cp -r ~/.cache/zig cache_zig_step0

echo -n "  Number of files in ~/.cache/zig snapshot cache_zig_step0: "
find cache_zig_step0 -type f | wc -l

echo

echo "ZIG_LIB_DIR build: expecting quick, seeing slow build"
_time env ZIG_LIB_DIR=s/lib/zig ${ZIGCC[@]} -DX=2 main.c -o /tmp/main2
cp -r ~/.cache/zig cache_zig_step1

echo -n "  Number of files in ~/.cache/zig snapshot cache_zig_step1: "
find cache_zig_step1 -type f | wc -l

echo
echo "To see files in cache_zig_step1 but not in cache_zig_step0, run:"
echo "  comm -3 <(cd cache_zig_step0; find -type f | sort) <(cd cache_zig_step1; find -type f | sort)"

```

Here is the output of the repro script. Take a particular notice to "Wall time":

```
Cleaning up and preparing directories... done
Warming up zig cache... done
Plain build: expecting a quick build
  + s/bin/zig cc -target x86_64-linux-gnu.2.28 -DX=1 main.c -o /tmp/main
  Wall time: 0.03
  Number of files in ~/.cache/zig snapshot cache_zig_stage0: 836

ZIG_LIB_DIR build: expecting quick, seeing slow build
  + env ZIG_LIB_DIR=s/lib/zig s/bin/zig cc -target x86_64-linux-gnu.2.28 -DX=2 main.c -o /tmp/main
  Wall time: 0.45
  Number of files in ~/.cache/zig snapshot cache_zig_stage1: 930

To see files in cache_zig_stage1 but not in cache_zig_stage0, run:
  comm -3 <(cd cache_zig_step0; find -type f | sort) <(cd cache_zig_step1; find -type f | sort)
```

<details><summary>click for the difference between ~/.cache/zig before and after the second compilation</summary>

```
        ./h/0ef1e1c08346b6a83d47a3076d461f1f.txt
        ./h/110d3b0fd6fb17bc3e2e317bb0b52bf2.txt
        ./h/1271630f1c451766c4e1a9d94ac132e5.txt
        ./h/128531048c4cd9c66a9e3d81176b6fa6.txt
        ./h/156d47066e29c7c82c9145685c75d908.txt
        ./h/15828d61727b58ee624be6d6897b6e3f.txt
        ./h/1608e7b082260dc9f285b5a63087d5fa.txt
        ./h/1827173279d4f78f0b76f6c00c6131cc.txt
        ./h/1a8f64fff5f1148088a5fbc1e708f706.txt
        ./h/204195480507ba49e55792bd606dadf9.txt
        ./h/2b2a1a6e19c3ee8c597f4faa46522129.txt
        ./h/3bb5acd92163c3220e1004664f6bdada.txt
        ./h/3c44cf3c4ed49dc57733cedbd3f7fb17.txt
        ./h/4153369738d015392ddacd1f72c75cd3.txt
        ./h/4a79990708eec7caaf1ca6c025ece2d8.txt
        ./h/4c08618fcc5a27c2ec7cd4058aa7347e.txt
        ./h/56e12a27edf73df27dc90e693ca95c24.txt
        ./h/5c6e4a4d0f94ad232809a49c0fe26694.txt
        ./h/61843fce1fb702eb177c9022bbf14510.txt
        ./h/67259e29a3d976b68d8def46bf32b73d.txt
        ./h/6b2ab327e64d66668a5b263dc108a817.txt
        ./h/777d38226e90d00ed4fe4b8897ccb653.txt
        ./h/7a6fcdebd2a78dc7cf31ea43d8c9e74c.txt
        ./h/7fd438690948ce224c6be5dfaca3eb2e.txt
        ./h/89db5980e4aea01f2d29ae63813117f9.txt
        ./h/8c34de7dc36ffe17bd21706ae67ca32f.txt
        ./h/90cbe014a1f6165884b1cf0a6fe297d3.txt
        ./h/92f57e1dcbc245f77adba80daa60e955.txt
        ./h/95f81993ad122be11b6c214c556f6d4d.txt
        ./h/9f553a88dce9319553f5b3acfdc292a5.txt
        ./h/aa7a6191b2af8bd9be0c597e8b4a7a5f.txt
        ./h/c1a21d346bead5475fc2670c40cd0978.txt
        ./h/c1e06d5beacebe82b0fda52ab288e45d.txt
        ./h/d002eba2587ac1bf9764b794d1869d4e.txt
        ./h/d360bce25c5191d2e41ec6d2c5c60fb5.txt
        ./h/d4abfe6610fd7e9cf9861c98846484f6.txt
        ./h/d55e9666fbf056385c0dd3403ab2f8d9.txt
        ./h/dca16b206ba1b8c78e26159ab9ec5ca9.txt
        ./h/e84f98a84bcf4289793a1b8d489b4bd8.txt
        ./h/f14502d9ee42fb6fb5bf6f5b29d80b81.txt
        ./h/f9b4bb4fc63ff4c90a21383e3a310bf1.txt
        ./h/fb1fa3b2816f27116c54f0b3026fa872.txt
        ./h/fbf0c8074a263abb86ebdcbf010be011.txt
        ./o/0257291e65b73163cb5684ceac63629c/util.o
        ./o/2e7c52edddf0086b1076abde7d9e237d/abi-note.o
        ./o/3143d20ce473f4c122fc243960b9819b/crti.o
        ./o/35b853e2bf268963d4e8261ec43adc8c/rt.o
        ./o/3b90808d7f7f18004735ce287976f587/crti.o
        ./o/41421eb9ec686cb5397eac799305733d/all.map
        ./o/41421eb9ec686cb5397eac799305733d/c.s
        ./o/41421eb9ec686cb5397eac799305733d/dl.s
        ./o/41421eb9ec686cb5397eac799305733d/ld.s
        ./o/41421eb9ec686cb5397eac799305733d/libc.so.6
        ./o/41421eb9ec686cb5397eac799305733d/libdl.so.2
        ./o/41421eb9ec686cb5397eac799305733d/libld.so.2
        ./o/41421eb9ec686cb5397eac799305733d/libm.so.6
        ./o/41421eb9ec686cb5397eac799305733d/libpthread.so.0
        ./o/41421eb9ec686cb5397eac799305733d/libresolv.so.2
        ./o/41421eb9ec686cb5397eac799305733d/librt.so.1
        ./o/41421eb9ec686cb5397eac799305733d/libutil.so.1
        ./o/41421eb9ec686cb5397eac799305733d/m.s
        ./o/41421eb9ec686cb5397eac799305733d/pthread.s
        ./o/41421eb9ec686cb5397eac799305733d/resolv.s
        ./o/41421eb9ec686cb5397eac799305733d/rt.s
        ./o/41421eb9ec686cb5397eac799305733d/util.s
        ./o/476911b0c368bb2228201f970e6a3f3c/at_quick_exit.o
        ./o/4aea22309715bc5b7405771ca97804ab/crtn.o
        ./o/52547f865830d9c6356fb4e5b052cdf4/fstatat.o
        ./o/5fe64761147ccf73c30ff22ffb667d3e/stat_t64_cp.o
        ./o/694df5124d282041fd54cb445346547c/m.o
        ./o/75cabd10fe73edc5428e63e262e7abfb/atexit.o
        ./o/805c8e15ebf71bbab0aa16cdcdb7dba1/dl.o
        ./o/8144f60f10af203bc9ba225ec51c2b8f/stat64.o
        ./o/81df8615cd5add00f556448f3e9e1809/Scrt1.o
        ./o/837a10b6e49d2ea0933dfc0005b498f7/lstat.o
        ./o/870289b1f1d2796a6e5a236eb742489d/pthread.o
        ./o/8956c8731d10312cbc32b5ad79f3f886/mknod.o
        ./o/91c8d1e169f58b5e67723694fc65d44c/crtn.o
        ./o/9703fb856ee48edcdc894cfcfd464e6a/stat.o
        ./o/a2af8b8d7008429fee04018d4e8a7025/elf-init-2.33.o
        ./o/a2d0cf009a3442263fbb918de395863e/fstat64.o
        ./o/a39ef2835c8e5bafe19eef3e8c9d682c/resolv.o
        ./o/a59386f3b1272d46d9dde2c95aa0d115/main.o
        ./o/a73ed027ced97e0e34afa29770068c1a/pthread_atfork.o
        ./o/ab2adf1787be8756e45f43c01b91a046/c.o
        ./o/abb9e825b719ba20560885be06db30a3/lstat64.o
        ./o/ba107fe8e5082422f28b9ab46a74c75f/start-2.33.o
        ./o/c0cf431d850a24bf6a7df99e77be9eb3/fstat.o
        ./o/cf5a81af0d8ddf0a4f00ea4162954c2b/libc_nonshared.a
        ./o/d08360a8dbdbb8704c35b14015692733/stack_chk_fail_local.o
        ./o/d7262a513e48316fafaf9ae87a71d987/ld.o
        ./o/dc619e8fec314e4bf96704cfa999ae5f/mknodat.o
        ./o/e2f77f0a41f388239aaf37805b02c28b/errno.o
        ./o/f3d63f95c6d4a9e31d2893e2de88744a/fstatat64.o
```

</details>

If I replace `cp -sr` with plain `cp -r` (i.e. copy real files, not symlinks), the issue does not manifest.

This message was written using `zig 0.11.0-dev.423+c37afa281` and I achieved similar results with #13805.

**Discussion**

The wall clock difference for a small compilation is about 10x and extra ~1.5MB to the global zig cache directory. Due to the nature how Bazel works, we pay the price for each linking step (which, in our environment, is the hundreds of thousands for a single test run).

We should find a way for zig to perform as well in either case: when glibc stubs are compiled from a "different" (according to zig's caching system) lib dir.

**Comment by motiejus** [comment]
Another data point:

```
diff --git a/src/Cache.zig b/src/Cache.zig
index 2c3213184..b80084247 100644
--- a/src/Cache.zig
+++ b/src/Cache.zig
@@ -493,9 +493,9 @@ pub const Manifest = struct {
                 self.failed_file_index = idx;
                 return err;
             };
-            const size_match = actual_stat.size == cache_hash_file.stat.size;
-            const mtime_match = actual_stat.mtime == cache_hash_file.stat.mtime;
-            const inode_match = actual_stat.inode == cache_hash_file.stat.inode;
+            const size_match = true; //actual_stat.size == cache_hash_file.stat.size;
+            const mtime_match = true; //actual_stat.mtime == cache_hash_file.stat.mtime;
+            const inode_match = true; //actual_stat.inode == cache_hash_file.stat.inode;
```

Does not fix the performance of the "naked" build as documented in [test-zigcc](https://git.sr.ht/~motiejus/test-zigcc#steps-to-reproduce). It still takes about the same time (150s in my last test compared to 142s in the README above).

So there is something else happening which causes zig to do lots of IO.

**Comment by motiejus** [comment]
io.stat without sandbox (17 second run):

```
253:1 rbytes=21767168 wbytes=94208 rios=619 wios=2 dbytes=0 dios=0
259:0 rbytes=497116160 wbytes=0 rios=10346 wios=0 dbytes=0 dios=0
253:0 rbytes=496718848 wbytes=16789504 rios=9799 wios=160 dbytes=0 dios=0
253:2 rbytes=474951680 wbytes=16695296 rios=9180 wios=158 dbytes=0 dios=0
```

With sandbox (140 second run):

```
253:1 rbytes=21803520 wbytes=143360 rios=614 wios=5 dbytes=0 dios=0
259:0 rbytes=497209856 wbytes=0 rios=10342 wios=0 dbytes=0 dios=0
253:0 rbytes=496845312 wbytes=149028864 rios=9803 wios=2197 dbytes=0 dios=0
253:2 rbytes=475041792 wbytes=148885504 rios=9189 wios=2191 dbytes=0 dios=0
```

The IO difference is insignificant, but time is. I will keep looking into this.

**Comment by rohlem** [comment]
@motiejus
> The IO difference is insignificant, but time is.
You may have missed it, `wbytes` actually gained a digit there!
In the third line `253:0` 16,789,504 without sandbox vs 149,028,864 with sandbox - with 1M bytes per second that rather linearly scales with the execution time as well.

**Comment by motiejus** [comment]
> @motiejus
> 
> > The IO difference is insignificant, but time is.
> 
> Not sure if I'm understanding the numbers wrong, or you may have missed it, `wbytes` actually gained a digit there! 

I was hyper focused on the reads. I indeed missed and now have a lead again. Thanks!!

**Comment by motiejus** [comment]
More data: I've again ran test-zigcc and measured the wall clock times of zig itself and bazel run:

| sandbox | nproc | bazel time | zig-cc time | rbytes | wbytes | rios | wios |
| --- | --- |--- | --- | --- | --- | --- | --- |
| yes | 8 |63s | 56s | 454410240 | 92397568 | 9235 |1065 |
| no | 8 | 15s | 31.5s | 453984256 | 16510976 | 9210 | 143 |
| yes | 1 | 95s | 21.5s | 452296704 | 37003264 | 9103 | 336 |
| no | 1 | 26s | 13.2s | 452804608 | 16265216 | 9135 | 139 |

Giving `zig cc` absolute and relative directories to `zig_lib_dir` makes no difference.

The dramatic drop in zig's total clock time between `nproc=8` and `nproc=1` seems to suggest there may be contention issue, as the synthetic workload is purely single-threaded (compile an object file and link it).

I can't yet explain the difference of `wbytes` between the sandboxed and non-sandboxed runs though.

**Comment by motiejus** [comment]
I took out the Bazel part and left only plain zig.

**Comment by motiejus** [comment]
For the record, 3ae4931dc1a3b1b338d2fd3a49a5a79b445bebf6 also exhibits the problem.

**Comment by motiejus** [comment]
Here is an observation: when `ZIG_LIB_DIR` is set to a symlink farm, it still creates the glibc stubs. However, it does it much faster than from scratch. Here is a script to reproduce the issue.

<details><summary>click for repro.sh</summary>

```bash
#!/bin/bash
set -euo pipefail

readonly ORIG=/code/zig/stage3
readonly ZIGCC=(s/bin/zig cc -target x86_64-linux-gnu.2.28)

_time() {
    /usr/bin/time --format "  + %C\n  Wall time: %e" "$@"
}

echo -n "Cleaning up and preparing directories... "
rm -fr s ~/.cache/zig cache_zig_step*
cp -sr $ORIG s
echo done

echo -n "Warming up zig cache... "
echo | ${ZIGCC[@]} /dev/stdin >& /dev/null || :
echo done

echo "Plain build: expecting a quick build"
_time ${ZIGCC[@]} -DX=1 main.c -o /tmp/main1
cp -r ~/.cache/zig cache_zig_step0

echo -n "  Number of files in ~/.cache/zig snapshot cache_zig_step0: "
find cache_zig_step0 -type f | wc -l

echo

echo "ZIG_LIB_DIR build: expecting quick, seeing slow build"
_time env ZIG_LIB_DIR=s/lib/zig ${ZIGCC[@]} -DX=2 main.c -o /tmp/main2
cp -r ~/.cache/zig cache_zig_step1

echo -n "  Number of files in ~/.cache/zig snapshot cache_zig_step1: "
find cache_zig_step1 -type f | wc -l

echo
echo "To see files in cache_zig_step1 but not in cache_zig_step0, run:"
echo "  comm -3 <(cd cache_zig_step0; find -type f | sort) <(cd cache_zig_step1; find -type f | sort)"

```
</details>

What this does:
1. Copy the whole zig dir as a "farm of symlinks": `cp -sr /code/zig/stage3 s`
2. Compile a file using `s/bin/zig cc <...>`.
3. Compile a file using `ZIG_LIB_DIR=s/lib/zig s/bin/zig cc <...>`
4. Time difference between (2) and (3) is about 10x.

Here is the output of the repro script. Take a particular notice to "Wall time":

```
Cleaning up and preparing directories... done
Warming up zig cache... done
Plain build: expecting a quick build
  + s/bin/zig cc -target x86_64-linux-gnu.2.28 -DX=1 main.c -o /tmp/main
  Wall time: 0.03
  Number of files in ~/.cache/zig snapshot cache_zig_stage0: 836

ZIG_LIB_DIR build: expecting quick, seeing slow build
  + env ZIG_LIB_DIR=s/lib/zig s/bin/zig cc -target x86_64-linux-gnu.2.28 -DX=2 main.c -o /tmp/main
  Wall time: 0.45
  Number of files in ~/.cache/zig snapshot cache_zig_stage1: 930

To see files in cache_zig_stage1 but not in cache_zig_stage0, run:
  comm -3 <(cd cache_zig_step0; find -type f | sort) <(cd cache_zig_step1; find -type f | sort)
```

<details><summary>click for the difference between ~/.cache/zig before and after the second compilation</summary>

```
        ./h/0ef1e1c08346b6a83d47a3076d461f1f.txt
        ./h/110d3b0fd6fb17bc3e2e317bb0b52bf2.txt
        ./h/1271630f1c451766c4e1a9d94ac132e5.txt
        ./h/128531048c4cd9c66a9e3d81176b6fa6.txt
        ./h/156d47066e29c7c82c9145685c75d908.txt
        ./h/15828d61727b58ee624be6d6897b6e3f.txt
        ./h/1608e7b082260dc9f285b5a63087d5fa.txt
        ./h/1827173279d4f78f0b76f6c00c6131cc.txt
        ./h/1a8f64fff5f1148088a5fbc1e708f706.txt
        ./h/204195480507ba49e55792bd606dadf9.txt
        ./h/2b2a1a6e19c3ee8c597f4faa46522129.txt
        ./h/3bb5acd92163c3220e1004664f6bdada.txt
        ./h/3c44cf3c4ed49dc57733cedbd3f7fb17.txt
        ./h/4153369738d015392ddacd1f72c75cd3.txt
        ./h/4a79990708eec7caaf1ca6c025ece2d8.txt
        ./h/4c08618fcc5a27c2ec7cd4058aa7347e.txt
        ./h/56e12a27edf73df27dc90e693ca95c24.txt
        ./h/5c6e4a4d0f94ad232809a49c0fe26694.txt
        ./h/61843fce1fb702eb177c9022bbf14510.txt
        ./h/67259e29a3d976b68d8def46bf32b73d.txt
        ./h/6b2ab327e64d66668a5b263dc108a817.txt
        ./h/777d38226e90d00ed4fe4b8897ccb653.txt
        ./h/7a6fcdebd2a78dc7cf31ea43d8c9e74c.txt
        ./h/7fd438690948ce224c6be5dfaca3eb2e.txt
        ./h/89db5980e4aea01f2d29ae63813117f9.txt
        ./h/8c34de7dc36ffe17bd21706ae67ca32f.txt
        ./h/90cbe014a1f6165884b1cf0a6fe297d3.txt
        ./h/92f57e1dcbc245f77adba80daa60e955.txt
        ./h/95f81993ad122be11b6c214c556f6d4d.txt
        ./h/9f553a88dce9319553f5b3acfdc292a5.txt
        ./h/aa7a6191b2af8bd9be0c597e8b4a7a5f.txt
        ./h/c1a21d346bead5475fc2670c40cd0978.txt
        ./h/c1e06d5beacebe82b0fda52ab288e45d.txt
        ./h/d002eba2587ac1bf9764b794d1869d4e.txt
        ./h/d360bce25c5191d2e41ec6d2c5c60fb5.txt
        ./h/d4abfe6610fd7e9cf9861c98846484f6.txt
        ./h/d55e9666fbf056385c0dd3403ab2f8d9.txt
        ./h/dca16b206ba1b8c78e26159ab9ec5ca9.txt
        ./h/e84f98a84bcf4289793a1b8d489b4bd8.txt
        ./h/f14502d9ee42fb6fb5bf6f5b29d80b81.txt
        ./h/f9b4bb4fc63ff4c90a21383e3a310bf1.txt
        ./h/fb1fa3b2816f27116c54f0b3026fa872.txt
        ./h/fbf0c8074a263abb86ebdcbf010be011.txt
        ./o/0257291e65b73163cb5684ceac63629c/util.o
        ./o/2e7c52edddf0086b1076abde7d9e237d/abi-note.o
        ./o/3143d20ce473f4c122fc243960b9819b/crti.o
        ./o/35b853e2bf268963d4e8261ec43adc8c/rt.o
        ./o/3b90808d7f7f18004735ce287976f587/crti.o
        ./o/41421eb9ec686cb5397eac799305733d/all.map
        ./o/41421eb9ec686cb5397eac799305733d/c.s
        ./o/41421eb9ec686cb5397eac799305733d/dl.s
        ./o/41421eb9ec686cb5397eac799305733d/ld.s
        ./o/41421eb9ec686cb5397eac799305733d/libc.so.6
        ./o/41421eb9ec686cb5397eac799305733d/libdl.so.2
        ./o/41421eb9ec686cb5397eac799305733d/libld.so.2
        ./o/41421eb9ec686cb5397eac799305733d/libm.so.6
        ./o/41421eb9ec686cb5397eac799305733d/libpthread.so.0
        ./o/41421eb9ec686cb5397eac799305733d/libresolv.so.2
        ./o/41421eb9ec686cb5397eac799305733d/librt.so.1
        ./o/41421eb9ec686cb5397eac799305733d/libutil.so.1
        ./o/41421eb9ec686cb5397eac799305733d/m.s
        ./o/41421eb9ec686cb5397eac799305733d/pthread.s
        ./o/41421eb9ec686cb5397eac799305733d/resolv.s
        ./o/41421eb9ec686cb5397eac799305733d/rt.s
        ./o/41421eb9ec686cb5397eac799305733d/util.s
        ./o/476911b0c368bb2228201f970e6a3f3c/at_quick_exit.o
        ./o/4aea22309715bc5b7405771ca97804ab/crtn.o
        ./o/52547f865830d9c6356fb4e5b052cdf4/fstatat.o
        ./o/5fe64761147ccf73c30ff22ffb667d3e/stat_t64_cp.o
        ./o/694df5124d282041fd54cb445346547c/m.o
        ./o/75cabd10fe73edc5428e63e262e7abfb/atexit.o
        ./o/805c8e15ebf71bbab0aa16cdcdb7dba1/dl.o
        ./o/8144f60f10af203bc9ba225ec51c2b8f/stat64.o
        ./o/81df8615cd5add00f556448f3e9e1809/Scrt1.o
        ./o/837a10b6e49d2ea0933dfc0005b498f7/lstat.o
        ./o/870289b1f1d2796a6e5a236eb742489d/pthread.o
        ./o/8956c8731d10312cbc32b5ad79f3f886/mknod.o
        ./o/91c8d1e169f58b5e67723694fc65d44c/crtn.o
        ./o/9703fb856ee48edcdc894cfcfd464e6a/stat.o
        ./o/a2af8b8d7008429fee04018d4e8a7025/elf-init-2.33.o
        ./o/a2d0cf009a3442263fbb918de395863e/fstat64.o
        ./o/a39ef2835c8e5bafe19eef3e8c9d682c/resolv.o
        ./o/a59386f3b1272d46d9dde2c95aa0d115/main.o
        ./o/a73ed027ced97e0e34afa29770068c1a/pthread_atfork.o
        ./o/ab2adf1787be8756e45f43c01b91a046/c.o
        ./o/abb9e825b719ba20560885be06db30a3/lstat64.o
        ./o/ba107fe8e5082422f28b9ab46a74c75f/start-2.33.o
        ./o/c0cf431d850a24bf6a7df99e77be9eb3/fstat.o
        ./o/cf5a81af0d8ddf0a4f00ea4162954c2b/libc_nonshared.a
        ./o/d08360a8dbdbb8704c35b14015692733/stack_chk_fail_local.o
        ./o/d7262a513e48316fafaf9ae87a71d987/ld.o
        ./o/dc619e8fec314e4bf96704cfa999ae5f/mknodat.o
        ./o/e2f77f0a41f388239aaf37805b02c28b/errno.o
        ./o/f3d63f95c6d4a9e31d2893e2de88744a/fstatat64.o
```

</details>

**Comment by andrewrk** [comment]
Thank you for the details and investigation. Over the next few days I will evaluate this feature request and see if the use case can be satisfied in a way that does not compromise any other use cases.

**Comment by motiejus** [comment]
> As you noted here, the real benefit would come if we can prevent zig from rehashing all these files, only to learn they are exactly the same. However, this modification will help mitigate the problem in the meantime:
> 
>     * [Replace SipHash-1-3 with AEGIS128L MAC for the Cache #13805](https://github.com/ziglang/zig/pull/13805)
> 
> 
> Perhaps it would be enough to unblock progress on your end, while waiting to see if the use case can be solved even better with more changes to zig?

It does not help. I built zig with #13805, here are the results:

```
$ ./repro.sh 
Cleaning up and preparing directories... done
Warming up zig cache... done
Plain build: expecting a quick build
  + /code/zig/build/stage3/bin/zig cc -target x86_64-linux-gnu.2.28 -DX=1 main.c -o /tmp/main1
  Wall time: 0.27
  Number of files in ~/.cache/zig snapshot cache_zig_step0: 836

ZIG_LIB_DIR build: expecting quick, seeing slow build
  + env ZIG_LIB_DIR=s/lib/zig /code/zig/build/stage3/bin/zig cc -target x86_64-linux-gnu.2.28 -DX=2 main.c -o /tmp/main2
  Wall time: 2.89
  Number of files in ~/.cache/zig snapshot cache_zig_step1: 930

To see files in cache_zig_step1 but not in cache_zig_step0, run:
  comm -3 <(cd cache_zig_step0; find -type f | sort) <(cd cache_zig_step1; find -type f | sort)
```

As mentioned in https://github.com/ziglang/zig/issues/13619#issuecomment-1333444562, a visible problem is additional nearly 100 files (mostly glibc stubs and their dependencies) *for each compliation*.

**Comment by andrewrk** [comment]
Thank you for your patience on this one. After looking into it more, I was able to identify the root cause as a bug in Zig's caching system. The zig lib directory path name was incorrectly making its way into the cache namespace for some build artifacts, such as libc_nonshared.a, atexit.o, and others, causing them to be rebuilt due to being in completely different cache namespaces.

I have already started working on a fix and expect to have it landed in master branch by tomorrow.
